src/r_util.cpp

```diff
@@ -504,6 +506,37 @@ namespace rhost {
             return R_FalseValue;
         }
 
+        extern "C" SEXP save_project(SEXP id, SEXP project_name, SEXP dest_dir, SEXP temp_dir) {
+            auto blob_id = static_cast<blobs::blob_id>(Rf_asReal(id));
+            fs::path prj_name, d_dir, t_dir;
+
+            r_top_level_exec([&]() {
+                prj_name = fs::u8path(Rf_translateCharUTF8(STRING_ELT(project_name, 0)));
+                d_dir = fs::u8path(Rf_translateCharUTF8(STRING_ELT(dest_dir, 0)));
+                t_dir = fs::u8path(Rf_translateCharUTF8(STRING_ELT(temp_dir, 0)));
+            }, __FUNCTION__);
+
+            fs::path zip_file = t_dir / prj_name;
+            zip_file.replace_extension(".zip");
+
+            if (fs::exists(zip_file)) {
+                fs::remove(zip_file);
+            }
+
+            blobs::save_to_file(blob_id, zip_file);
+
+            fs::path temp_proj_dir = t_dir / prj_name;
+            if (fs::exists(temp_proj_dir)) {
+                fs::remove_all(temp_proj_dir);
+            }
+
+            fs::path dest_proj_dir = d_dir / prj_name;
+
+            rproj::extract_project(zip_file, dest_proj_dir, temp_proj_dir);
```

Comment:
Or maybe  not extract_project, since you also need to handle errors for save_to_file above. But extract it all to a separate function, e.g. save_to_project_folder_worker, and have save_to_project_folder call that from inside exceptions_to_errors.

---
In reply to: [74851442](https://github.com/Microsoft/R-Host/pull/78#discussion_r74851442) [](ancestors = 74851442,74672717)